src/optimizer/simple_optimizer.cpp

```diff
@@ -1023,5 +1025,153 @@ std::shared_ptr<const peloton::catalog::Schema> CreateHackJoinSchema() {
+  // Generate hash for right table
+  auto right_key = expression::ExpressionUtil::ConvertToTupleValueExpression(right_schema, right_key_col_name);
+  std::vector<std::unique_ptr<const expression::AbstractExpression>> hash_keys;
+  hash_keys.emplace_back(right_key);
+
+  // Create hash plan node
+  std::unique_ptr<planner::HashPlan> hash_plan_node(new planner::HashPlan(hash_keys));
+  hash_plan_node->AddChild(std::move(right_SelectPlan));
+
+  std::vector<catalog::Column> output_table_columns = {};
+  // expressions to evaluate
+  TargetList tl = TargetList();
+  // columns which can be returned directly
+  DirectMapList dml = DirectMapList();
+  auto &select_list = *select_stmt->getSelectList();
+  int i = 0;
+
+  // SELECT * FROM A JOIN B
+  if (select_list[0]->GetExpressionType()==EXPRESSION_TYPE_STAR) {
+	auto &left_cols = left_schema->GetColumns();
+	auto &right_cols = right_schema->GetColumns();
+	output_table_columns = left_cols;
+	output_table_columns.insert(output_table_columns.end(), right_cols.begin(), right_cols.end());
+	int left_col_cnt = left_cols.size();
+	int right_col_cnt = right_cols.size();
+	for (int j=0; j<left_col_cnt; j++) {
+		dml.push_back(DirectMap(i++, std::make_pair(0, j)));
+	}
+	for (int j=0; j<right_col_cnt; j++) {
+		dml.push_back(DirectMap(i++, std::make_pair(1, j)));
+	}
+  }
+
+  // SELECT col1, col2, fun1, ... FROM A JOIN B
+  else {
+	  for (auto expr : select_list) {
+		  auto expr_type = expr->GetExpressionType();
+		  // Column experssion
+		  if (expr_type == EXPRESSION_TYPE_VALUE_TUPLE) {
+			  auto tup_expr = (expression::TupleValueExpression*) expr;
+			  oid_t old_col_id = left_schema->GetColumnID(tup_expr->GetColumnName());
+			  catalog::Column column;
+			  // The current column is in the left table
+			  if (old_col_id!=(oid_t)-1) {
+				  column = left_schema->GetColumn(old_col_id);
+				  dml.push_back(DirectMap(i, std::make_pair(0, old_col_id)));
+			  }
+			  // The current column is in the right table
+			  else {
+				  old_col_id = right_schema->GetColumnID(tup_expr->GetColumnName());
+				  column = right_schema->GetColumn(old_col_id);
+				  dml.push_back(DirectMap(i, std::make_pair(1, old_col_id)));
+			  }
+			  output_table_columns.push_back(column);
+
+
+
+			auto type = column.GetType();
+			// set the expression name to the alias if we have one
+			if (tup_expr->alias.size() > 0) {
+				tup_expr->expr_name_ = tup_expr->alias;
+			} else {
+				tup_expr->expr_name_ = tup_expr->GetColumnName();
+			}
+			// point to the correct column returned in the logical tuple underneath
+			tup_expr->SetTupleValueExpressionParams(type, i, 0);
+		  }
+
+		  // Function Ref
+		  else {
+			  tl.push_back(Target(i, expr->Copy()));
+			  output_table_columns.push_back(
+				  catalog::Column(expr->GetValueType(),
+					  common::Type::GetTypeSize(expr->GetValueType()),
+					  "expr" + std::to_string(i)));
+		  }
+		  i++;
+	  }
+  }
```

Comment:
Instead of another set of separate logic determining output columns and mapping list, we should reuse the logic here:
https://github.com/cmu-db/peloton/blob/master/src/include/expression/expression_util.h#L299

I know that utility function can only handle the single table case, so you probably have to refactor that logic a little bit. I think replacing the single schema reference in the parameter to a reference of a vector of schemas would do. You also need to supplement the logic in that function (e.g., get output columns for star expression).

I think if we bite the bullet to refactor that utility function right now, it will also save us lots of time later. You can talk to me if you're not sure what you should to do.